docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,137 @@
+
+#### Class
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getName()` | `table` | returns the name of self's class |
+| `getParent()` | `lType` | returns lType userdata of self's parent |
+
+</details>
+
+### Implementation
+
+The implementation of user-defined type functions can be broken down into the following parts:
+
+**AST**: Introduce a new AST node for type functions called `AstStatTypeFunction`.
+
+**Parser**: When parsing TypeAliases, if the next lexeme is "function", parse it into the `AstStatTypeFunction` node.
+
+**Constraint Generator**: Generate a `TypeAliasExpansionConstraint` when visiting the `AstStatTypeFunction` node.
+
+**Constraint Solver**: Generate a new `ReduceConstraint` and append it to the list of unsolved constraints when reducing `TypeAliasExpansionConstraint`.
+
+**Reduction Constraint**: Serialize the function arguments into `lType` and execute the body of the function. Reduce to the deserialized version of the function return value.
+
+**lType**: Using the Lua API, create a library that interfaces between C++ and Luau and supports all of the function calls, serialization, and deserialization.
+
+## Drawback
+
+Type functions are handled at the analysis time, while `lType` is an implementation in the runtime. As a result, the proposed design causes the analysis time to be dependent on Luau's runtime to reduce user-defined type functions. This is generally discouraged as it is best to isolate the compile time, analysis time, and runtime from each other for the purpose of maintaining a clean separation of concerns, which helps minimize side effects and dependencies across different phases of the program execution and improves the modularity of the compiler. Overlaps between the analysis time and runtime can lead to code that is more complex and harder to manage, potentially increasing the risk of bugs and making the outcomes less predictable.
+
+The build / analysis times will also be negatively impacted by this feature as reducing type functions takes variable amount of time based on the program. The larger the type function, the longer it will take to reduce it in the constraint solver.
+
+## Alternatives
+
+### `table` Runtime Representation
+
+Currently, the runtime representation of types is a userdata called `lType`. Another representation is to use the already-existing type in Luau `table`; instead of serializing types into `lType`, we can serialize them into tables with predefined properties. For instance, the representation for a string singleton `"abc"` could be `{type = "stringSingleton", value = "abc"}`. So instead of writing:
+```luau
+type function isSingleton(t)
+    if t:isStringSingleton() then
+        return t
+    end
+end
+```
+developers could write:
+```luau
+type function isSingleton(t)
+    if t.type == "stringSingleton" then
+        return t
+    end
+end
+```
```

Comment:
Both of these type functions are missing a final return value for the failing cases.

Reply:
functions have implied return value of nil, so my intention was to show that these type functions will fail to reduce if it returns nil / doesn't specify a return value. This was also mentioned in the "Design" section when I wrote an example for `rawget`.

Reply:
I was hoping we would use `error("reason")` instead of `return nil` or `return`. The error function would then get displayed in type errors.

Reply:
Functions do not have an implied return of `nil`. Functions that do not have a return statement do not return anything, which is distinct from returning a `nil` value. You can see this by looking at the semantics of `select("#", ...)` on a function that has `return nil` vs no return.